temporalcli/commands.schedule.go

```diff
@@ -0,0 +1,597 @@
+			{
+				Start:   c.StartTime.Time(),
+				End:     c.EndTime.Time(),
+				Overlap: overlap,
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+	cctx.Printer.Println("Backfill request sent")
+	return nil
+}
+
+func toCronString(pb *schedpb.CalendarSpec) (string, error) {
+	def := func(a, b string) string {
+		if a != "" {
+			return a
+		}
+		return b
+	}
+	fields := []string{
+		def(pb.Second, "0"),
+		def(pb.Minute, "0"),
+		def(pb.Hour, "0"),
+		def(pb.DayOfMonth, "*"),
+		def(pb.Month, "*"),
+		def(pb.DayOfWeek, "*"),
+		def(pb.Year, "*"),
+	}
+	for _, f := range fields {
+		if len(strings.Fields(f)) != 1 {
+			return "", fmt.Errorf("invalid CalendarSpec")
+		}
+	}
+	if pb.Comment != "" {
+		fields = append(fields, "#", pb.Comment)
+	}
+	return strings.Join(fields, " "), nil
+}
+
+func toIntervalSpec(str string) (client.ScheduleIntervalSpec, error) {
+	var spec client.ScheduleIntervalSpec
+	var err error
+	parts := strings.Split(str, "/")
+	if len(parts) > 2 {
+		return spec, fmt.Errorf(`invalid interval: must be "<duration>" or "<duration>/<duration>"`)
+	} else if len(parts) == 2 {
+		if spec.Offset, err = timestamp.ParseDuration(parts[1]); err != nil {
+			return spec, fmt.Errorf("invalid interval: %w", err)
+		}
+	}
+	if spec.Every, err = timestamp.ParseDuration(parts[0]); err != nil {
+		return spec, fmt.Errorf("invalid interval: %w", err)
+	}
+	return spec, nil
+}
+
+func (c *ScheduleConfigurationOptions) toScheduleSpec(spec *client.ScheduleSpec) error {
+	spec.CronExpressions = c.Cron
+	// Skip not supported
+	spec.Jitter = c.Jitter
+	spec.TimeZoneName = c.TimeZone
+	spec.StartAt = c.StartTime.Time()
+	spec.EndAt = c.EndTime.Time()
+
+	var err error
+	for _, calPbStr := range c.Calendar {
+		var calPb schedpb.CalendarSpec
+		if err = protojson.Unmarshal([]byte(calPbStr), &calPb); err != nil {
+			return fmt.Errorf("failed to parse json calendar spec: %w", err)
+		}
+		cron, err := toCronString(&calPb)
+		if err != nil {
+			return err
+		}
+		spec.CronExpressions = append(spec.CronExpressions, cron)
+	}
+	for _, intStr := range c.Interval {
+		int, err := toIntervalSpec(intStr)
+		if err != nil {
+			return err
+		}
+		spec.Intervals = append(spec.Intervals, int)
+	}
+
+	return nil
+}
+
+func toScheduleAction(sw *SharedWorkflowStartOptions, i *PayloadInputOptions) (client.ScheduleAction, error) {
+	opts, err := buildStartOptions(sw, &WorkflowStartOptions{})
+	if err != nil {
+		return nil, err
+	}
+	action := &client.ScheduleWorkflowAction{
+		ID:                       opts.ID,
+		Workflow:                 sw.Type,
+		TaskQueue:                opts.TaskQueue,
+		WorkflowExecutionTimeout: opts.WorkflowExecutionTimeout,
+		WorkflowRunTimeout:       opts.WorkflowRunTimeout,
+		WorkflowTaskTimeout:      opts.WorkflowTaskTimeout,
+		// RetryPolicy not supported yet
+		// TODO: buildStartOptions doesn't fill in TypedSearchAttributes, this doesn't work yet:
+		TypedSearchAttributes: opts.TypedSearchAttributes,
```

Comment:
@cretz @Quinn-With-Two-Ns What do you think is the best way to handle this right now?

The situation appears to be: `buildStartOptions` builds search attributes in the form `map[string]any` (very easy to do from json) and puts them in `StartWorkflowOptions.SearchAttributes`. This is deprecated, but it still works for starting workflows.

For `ScheduleWorkflowAction` though, apparently untyped support is gone completely? The comment on `UntypedSearchAttributes` says `This should never be used for create.` Even if I could use it, it's a `map[string]*commonpb.Payload` which means I'd have to convert to `Payload`s (a little more work but not a huge deal).

To set `TypedSearchAttributes`, though, it looks like I'd have to do a lot of work. Basically it would have to look a lot like `convertToTypedSearchAttributes` from `internal/internal_search_attributes.go`, except with reflection?

I understand the goal of typed search attributes is to make them hard to use in a generic way, but the cli kinda needs to use them in a generic way 🫤  (So does "modify schedule search attributes" btw, except that's from workflow context so the code has to be different again.)

Reply:
Ug, yeah, we've hit this a few times now with generic tools. If the `UntypedSearchAttributes` does work for now I think it's ok to use here and just assume that warning is for users only. The problem is technically a search attribute key for setting is supposed to be defined by its name _and_ type, but we have no reasonable interface for that in CLI.

Note the starting of workflows sets untyped search attributes too, we were just forced to remove the either-or option on schedules because this same type is used for update and we couldn't tell which was the intended field on update (it was a long discussion internally before we made this backwards compatibility break and we documented it in release notes).

Reply:
I confirmed `UntypedSearchAttributes` in the start workflow action works for create schedule so I used that for now (sorry Quinn) and added a test.